apps/typegpu-docs/src/content/docs/integration/webgpu-interoperability.mdx

```diff
@@ -0,0 +1,327 @@
+
+Since TypeGPU is a very thin abstraction over WebGPU, there is a 1-to-1 mapping between most typed resources
+and the raw WebGPU resource used underneath.
+
+```ts
+const layout = tgpu.bindGroupLayout(...);
+//    ^? TgpuBindGroupLayout<...>
+
+const rawLayout = root.unwrap(layout); // => GPUBindGroupLayout
+```
+
+```ts
+const numbersBuffer = root.createBuffer(d.f32).$usage('uniform');
+//    ^? TgpuBuffer<d.F32> & Uniform
+
+const rawNumbersBuffer = root.unwrap(numbersBuffer); // => GPUBuffer
+
+// Operations on `rawNumbersBuffer` and `numbersBuffer` are shared, because
+// they are essentially the same resource.
+```
+
+```ts
+const fooTexture = root.createTexture({ ... }).$usage('sampled');
+//    ^? TgpuTexture<...> & Sampled
+
+const rawFooTexture = root.unwrap(fooTexture); // => GPUTexture
+```
+
+```ts
+const bindGroup = root.createBindGroup(layout, { ... });
+//    ^? TgpuBindGroup<...>
+
+const rawBindGroup = root.unwrap(bindGroup); // => GPUBindGroup
+```
+
+:::tip
+Each call to `root.unwrap(...)` with the same argument produces the exact same result, so there is no need to
+store the raw equivalents anywhere. Just call `root.unwrap(...)` in multiple places for the same resource if
+it makes the code easier to work with.
+:::
+
+
+### Plugging WebGPU resources into typed APIs.
+
+Many TypeGPU APIs accept either typed resources, or their untyped equivalent.
+
+#### Buffers
+
+Instead of passing an initial value to `root.createBuffer`, we can pass it a raw WebGPU buffer and interact with it through
+TypeGPU's APIs.
+
+```ts
+const rawBuffer = device.createBuffer({
+  size: (Float32Array.BYTES_PER_ELEMENT * 4) * 2, // (xyz + padding) * 2
+  usage: GPUBufferUsage.COPT_DST | GPUBufferUsage.UNIFORM,
+});
+
+const Schema = d.arrayOf(d.vec3f, 2);
+const aBuffer = root.createBuffer(Schema, rawBuffer);
+
+// Updates `rawBuffer` underneath.
+aBuffer.write([d.vec3f(1, 2, 3), d.vec3f(4, 5, 6)]);
+
+// Interpreting the raw bytes in `rawBuffer` as easier
+// to work with JS values.
+const values = await aBuffer.read(); // => d.v3f[]
+```
+
+#### Bind Group Layouts
+
+When creating typed bind groups from a layout, entries can be populated with equivalent raw WebGPU resources.
+
+```ts
+const layout = tgpu.bindGroupLayout({
+  a: { uniform: d.f32 },
+  b: { uniform: d.u32 },
+});
+
+const aBuffer = root.createBuffer(d.f32, 0.5).$usage('uniform');
+//    ^? TgpuBuffer<d.F32> & Uniform
+
+const bBuffer = root.device.createBuffer({
+  size: Uint32Array.BYTES_PER_ELEMENT,
+  usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.UNIFORM,
+});
+// ^? GPUBuffer
+
+const bindGroup = root.createBindGroup(layout, {
+  a: aBuffer, // validates if the buffers holds the right data and usage on the type level.
+  b: bBuffer, // allows the raw buffer to pass through.
+});
+```
+
+## Incremental adoption recipes
+
+Let us deliver on our promise and show you the small code changes necessary to adopt TypeGPU, and the benefits you can gain at each step.
```

Comment:
Sound a little weird to me but I can't pinpoint why. Maybe a little too dialogue'y? Maybe something like `To deliver on the promise of interoperability ...`?